include/aws/crt/Variant.h

```diff
@@ -0,0 +1,348 @@
+                    return std::is_same<T, Last>::value ? curIndex : -1;
+                }
+
+                template<typename T, typename First, typename Second, typename ...Rest>
+                constexpr std::size_t GetIndexOf(std::size_t curIndex = 0) {
+                    return std::is_same<T, First>::value ? curIndex : GetIndexOf<T, Second, Rest...>(curIndex++);
+                }
+            }
+
+            namespace Checker
+            {
+                // Returns True if the template parameter pack Ts has a type T, i.e. ContainsType<T, Ts>() == true if T is in the list of Ts
+//                template<typename T, typename First, typename Second, typename ...Rest>
+//                constexpr bool ContainsType() {
+//                    return std::is_same<T, First>::value || ContainsType<T, Second, Rest...>();
+//                }
+                template<typename T, typename Last>
+                constexpr bool ContainsType() {
+                    return std::is_same<T, Last>::value;
+                }
+
+                template<typename T, typename First, typename Second, typename ...Rest>
+                constexpr bool ContainsType() {
+                    return std::is_same<T, First>::value || ContainsType<T, Second, Rest...>();
+                }
+
+//                template<typename T, typename ...Ts>
+//                constexpr bool ContainsType() {
+//                    return ContainsType<T, Ts...>();
+//                }
+
+
+
+                // a case when the template parameter pack is empty (i.e. Variant<>)
+                template<typename T>
+                constexpr bool ContainsType() {
+                    return false;
+                }
+
+                template<typename T, typename ...Ts>
+                struct HasType
+                {
+                    static const bool value = ContainsType<T, Ts...>();
+                };
+            }
+        }
+
+        template <std::size_t Index, typename... Ts> class VariantAlternative;
+
+        /**
+         * Custom implementation of a Variant type. std::variant requires C++17
+         * @tparam Ts types of the variant value
+         */
+        template <typename... Ts> class Variant
+        {
+          public:
+            static constexpr std::size_t AlternativeCount = sizeof...(Ts);
+
+            template<std::size_t Index>
+            using VariantAlternative = VariantAlternative<Index, Variant<Ts...>>;
+
+            using ThisVariantT = Variant<Ts...>;
+            template<typename OtherT>
+            using EnableIfOtherIsSameVariantType = typename std::enable_if<std::is_same<typename std::decay<OtherT>, ThisVariantT>::value, int>::type;
+
+            template<typename OtherT>
+            using EnableIfOtherIsThisVariantAlternative = typename std::enable_if<VariantDetail::Checker::HasType<typename std::decay<OtherT>::type, Ts...>::value, int>::type;
+
+            Variant() : m_value(nullptr) {}
```

Comment:
docs say:
> Constructs a variant holding the [value-initialized](https://en.cppreference.com/w/cpp/language/value_initialization) value of the first alternative ([index()](https://en.cppreference.com/w/cpp/utility/variant/index) is zero).

so we need to set m_index=0, and need to actually run a constructor